tests/aliquot/test_aliquot_models.py

```diff
@@ -0,0 +1,384 @@
+        #get Aliquot
+        a= Aliquot.query.filter_by(external_id=aliquot_id).one_or_none()
+        
+        #Delete Aliquot
+        db.session.delete(a)
+        db.session.commit()
+
+        a= Aliquot.query.filter_by(external_id=aliquot_id).one_or_none()
+        self.assertIs(a,None)
+        self.assertIsNone(Sample.query.filter_by(external_id=aliquot_id).one_or_none())
+        self.assertEqual(Sample.query.count(),1)
+        self.assertEqual(Participant.query.count(),1)
+
+
+    def test_delete_aliquot_sample_participant(self):
+        """
+        Test deleting aliquot via sample and participant
+        """
+
+        participant_id, sample_id, aliquot_id= self.create_participant_sample_aliquot()
+
+        # Delete Participant
+        p = Participant.query.first()
+        db.session.delete(p)
+        db.session.commit()
+
+        s= Sample.query.filter_by(external_id=sample_id).one_or_none()
+        self.assertIs(s,None)
+        # Check no sample exists in participant
+        self.assertIsNone(Participant.query.filter_by(external_id=sample_id).one_or_none())
+
+        a= Aliquot.query.filter_by(external_id=aliquot_id).one_or_none()
+        self.assertIs(a,None)
+
+        #Check No aliquot exists in sample
+        self.assertIsNone(Sample.query.filter_by(external_id=aliquot_id).one_or_none())
+
+    def test_not_null_constraint(self):
+        """
+        Test aliquot cannot be created with out sample or participant
+        """    
+        dt = datetime.now()
+        # Create aliquot without sample kf_id
+        aliquot_id = "Test_Aliquot_0"
+        aliquot_data = {
+                    'external_id':'Test_Aliquot_0',
+                    'shipment_origin':'CORIELL',
+                    'shipment_destination':'Broad Institute',
+                    'analyte_type':'DNA',
+                    'concentration':100,
+                    'volume':12.67,
+                    'shipment_date':dt
+                    }
+        a = Aliquot(**aliquot_data)
+        
+        #Add Aliquot to db
+        self.assertRaises(IntegrityError, db.session.add(a))
+
+    def test_foreign_key_constraint(self):
+
+        """
+        Test aliquot cannot be created with empty sample_id
+    
+        """  
+        dt = datetime.now()  
+        # Create aliquot
+        aliquot_id = "Test_Aliquot_0"
+        aliquot_data = {
+                    'external_id':'Test_Aliquot_0',
+                    'shipment_origin':'CORIELL',
+                    'shipment_destination':'Broad Institute',
+                    'analyte_type':'DNA',
+                    'concentration':100,
+                    'volume':12.67,
+                    'shipment_date':dt
+                    }
+        a = Aliquot(**aliquot_data,sample_id = '')
+
+        #Add aliquot to db
+        self.assertRaises(IntegrityError, db.session.add(a))
+
+    def test_one_to_many_relationship_create(self):
```

Comment:
Not sure that this is truly testing the one to many relationship between Sample and Aliquot. You are creating Sample s1 and adding Aliquot a1 to it. Then you are creating Sample s2 and adding aliquot a2 to it. I think you might want to try adding Aliquot a2 to Sample s1 so that you can verify that one sample has many aliquots.

Reply:
yes I could do that too. My initial thought was to test multi relationship between participant to sample and finally to aliquot. I will add another aliquot to one of the sample.